### PR TITLE
Implement "Delete Save File" Menu Item

### PR DESCRIPTION
**Pull request by Rstar284**
Ryujinx
 
Add a delete save file menu item when right clicking on a game

Known problems:

~~Device Saves not deleting.~~ Fixed

### CODE REVIEW

**Comment by GamerzHell9137** [comment]
Haven't checked the PR but if it doesn't have a confirmation window to delete the save, it should be implemeneted.
"Are you sure you want to delete this save file? Files can't be recovered once deleted! [YES] [NO] "or something like that

**Comment by Rstar284** [comment]
image

mhm! cant have such a dangerous action happen without conformation.

**Comment by Thealexbarney** [comment]
The way this is done is incorrect. You're not deleting the saves, you're just cleaning out the save directory. And you usually shouldn't interact with the file system using the built-in .NET APIs. You should use the `DeleteSaveData`, `DeleteDeviceSaveData`, etc. APIs instead via `_horizonClient.Fs` in the `GameTableContextMenu` object.

**Comment by Rstar284** [comment]
Will fix, thanks for that :)

**Comment by Thealexbarney** [comment]
You'd probably want deleting an account save and the device save to be separate actions, as well as deleting each account's save individually. You also aren't doing anything when the `Delete*` functions don't return `Success`. `.IsSuccess()` should be used when checking for a successful result.

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 1]
```diff
@@ -454,6 +455,29 @@ private void OpenSaveBcatDir_Clicked(object sender, EventArgs args)
             OpenSaveDir(saveDataFilter);
         }
 
+        private void PurgeSave_Clicked(object sender, EventArgs args)
+        {
+            SaveDataFilter saveDataFilter = new SaveDataFilter();
+            saveDataFilter.SetProgramId(new ProgramId(_titleId));
+            if (!TryFindSaveData(_titleName, _titleId, _controlData, saveDataFilter, out ulong saveDataId)) {return;}
```
Comment (riperiperi) [inline Ryujinx/Ui/Widgets/GameTableContextMenu.cs:462]: Should add newlines around the curly braces here to match our usual format for if blocks.

File: Ryujinx/Ui/Widgets/GameTableContextMenu.Designer.cs [hunk 2]
```diff
@@ -50,6 +51,15 @@ private void InitializeComponent()
                 TooltipText = "Open the directory which contains Application's BCAT Saves."
             };
             _openSaveBcatDirMenuItem.Activated += OpenSaveBcatDir_Clicked;
+            
+            //
+            // _purgeSaveMenuItem
+            //
+            _purgeSaveMenuItem = new MenuItem("Delete Save File")
+            {
+                TooltipText = "Delete all the Application's Save file (if they exists)."
```
Comment (riperiperi) [inline Ryujinx/Ui/Widgets/GameTableContextMenu.Designer.cs:60]: `Delete all save files for the Application (if they exist)`

**Review by riperiperi — CHANGES_REQUESTED** [review]
Just a few comments on style and wording.

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 3]
```diff
@@ -455,29 +457,58 @@ private void OpenSaveBcatDir_Clicked(object sender, EventArgs args)
...
-            MessageDialog warningDialog = GtkDialog.CreateConfirmationDialog("Warning", $"You are about to delete all savedata for :\n<b>{_titleName}</b>\nAre you sure you want to proceed?\nFiles can't be recovered once deleted!");
+            if (!TryFindSaveData(_titleName, _titleId, _controlData, saveDataFilter, out ulong saveDataId))
+            {
+                return;
+            }
+            MessageDialog warningDialog = GtkDialog.CreateConfirmationDialog("Warning", $"You are about to delete the {saveDataType.ToString().ToLower()}'s savedata for :\n<b>{_titleName}</b>\nAre you sure you want to proceed?\nFiles can't be recovered once deleted!");
             if (warningDialog.Run() == (int)ResponseType.Yes)
             {
-                try
+                if (saveDataType.Equals(SaveDataType.Account))
```
Comment (riperiperi) [inline Ryujinx/Ui/Widgets/GameTableContextMenu.cs:470]: Just use `==` when comparing an enum.

**Review by riperiperi — CHANGES_REQUESTED** [review]

File: Ryujinx/Ui/Widgets/GameTableContextMenu.Designer.cs [hunk 4]
```diff
@@ -53,13 +54,22 @@ private void InitializeComponent()
...
+            _purgeSaveUserMenuItem.Activated += PurgeUserSaveData_Clicked;
+            
+            //
+            // _purgeSaveDeviceMenuItem
+            //
+            _purgeSaveDeviceMenuItem = new MenuItem("Delete Device Save Data")
+            {
+                TooltipText = "Delete the Device's Application's Save file (if it exists)."
+            };
+            _purgeSaveUserMenuItem.Activated += PurgeDeviceSaveData_Clicked;
```
[the remark this answers is not in the document]
Comment (Rstar284) [reply Ryujinx/Ui/Widgets/GameTableContextMenu.Designer.cs:72]: Haha, i was thinking why both menus opened together

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 5]
```diff
@@ -454,6 +457,58 @@ private void OpenSaveBcatDir_Clicked(object sender, EventArgs args)
             OpenSaveDir(saveDataFilter);
         }
 
+        private void PurgeSaveData(SaveDataFilter saveDataFilter, SaveDataType saveDataType)
+        {
+            saveDataFilter.SetProgramId(new ProgramId(_titleId));
+            if (!TryFindSaveData(_titleName, _titleId, _controlData, saveDataFilter, out ulong saveDataId))
```
Comment (Thealexbarney) [inline Ryujinx/Ui/Widgets/GameTableContextMenu.cs:463]: `TryFindSaveData` appears to ask to create the save data if it doesn't already exist. That's not what you'd want in this situation. Also the filter you're passing in will check for any save with the given program ID, regardless of save type or user ID.

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 6]
```diff
@@ -454,6 +457,58 @@ private void OpenSaveBcatDir_Clicked(object sender, EventArgs args)
             OpenSaveDir(saveDataFilter);
         }
 
+        private void PurgeSaveData(SaveDataFilter saveDataFilter, SaveDataType saveDataType)
```
[the remark this answers is not in the document]
Comment (Rstar284) [reply Ryujinx/Ui/Widgets/GameTableContextMenu.cs:460]: Uh well, i just copied how the other delete function were called

File: Ryujinx/Ui/Widgets/GameTableContextMenu.cs [hunk 7]
```diff
@@ -68,6 +69,8 @@ public GameTableContextMenu(MainWindow parent, VirtualFileSystem virtualFileSyst
             _openSaveUserDirMenuItem.Sensitive   = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.UserAccountSaveDataSize      > 0;
             _openSaveDeviceDirMenuItem.Sensitive = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.DeviceSaveDataSize           > 0;
             _openSaveBcatDirMenuItem.Sensitive   = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.BcatDeliveryCacheStorageSize > 0;
+            _purgeSaveDeviceMenuItem.Sensitive   = !Utilities.IsZeros(controlData.ByteSpan) && controlData.Value.DeviceSaveDataSize           > 0;
```
[the remark this answers is not in the document]
Comment (Rstar284) [reply Ryujinx/Ui/Widgets/GameTableContextMenu.cs:72]: I dont see a way to do so, if you know could help me

**Comment by marysaka** [comment]
Closing as stale

**Pull request closed without merging**